src/java.base/share/classes/java/lang/invoke/VarHandles.java

```diff
@@ -710,10 +710,10 @@ private static boolean isCheckedException(Class<?> clazz) {
 //                @LambdaForm.Compiled
 //                @Hidden
 //                static final <METHOD> throws Throwable {
-//                    handle.checkExactAccessMode(ad);
-//                    if (handle.isDirect() && handle.vform.methodType_table[ad.type] == ad.symbolicMethodTypeErased) {
+//                    boolean direct = handle.checkExactAccessMode(ad);
+//                    if (direct && handle.vform.methodType_table[ad.type] == ad.symbolicMethodTypeErased) {
 //                        MethodHandle.linkToStatic(<LINK_TO_STATIC_ARGS>);
-//                    } else if (handle.isDirect() && handle.vform.getMethodType_V(ad.type) == ad.symbolicMethodTypeErased) {
+//                    } else if (direct && handle.vform.getMethodType_V(ad.type) == ad.symbolicMethodTypeErased) {
 //                        MethodHandle.linkToStatic(<LINK_TO_STATIC_ARGS>);
 //                    } else {
 //                        MethodHandle mh = handle.getMethodHandle(ad.mode);
```

Comment:
The `direct`‑ness check can be hoisted into an enclosing `if` statement:
```suggestion
//                    if (direct) {
//                        if (handle.vform.methodType_table[ad.type] == ad.symbolicMethodTypeErased) {
//                            MethodHandle.linkToStatic(<LINK_TO_STATIC_ARGS>);
//                            return;
//                        } else if (handle.vform.getMethodType_V(ad.type) == ad.symbolicMethodTypeErased) {
//                            MethodHandle.linkToStatic(<LINK_TO_STATIC_ARGS>);
//                            return;
//                        }
//                    }
//                    MethodHandle mh = handle.getMethodHandle(ad.mode);
```

Also, any reason `GUARD_METHOD_TEMPLATE_V` uses `LINK_TO_STATIC_ARGS` instead of `LINK_TO_STATIC_ARGS_V`?

Reply:
> How would the performance change if the `isDirect` and `checkExactAccessMode` merger was reverted?

Add around 15-20ns/op for these micros.

Reply:
Restructuring so that we only check `direct` once sounds reasonable at face value but would be a lot of churn for little gain (even in the interpreter testing a local boolean field is fast, and JITs will optimize this well).

The `LINK_TO_STATIC_ARGS_V` in the code generator seems like a remnant from an earlier iteration of this code. The `vform.getMemberName_V` method the code generator would emit a call to doesn't even exist. This should probably be cleaned up, separately. @PaulSandoz, WDYT?

Reply:
I think it was some left-over from prototyping and was not cleaned up. AFAICT `getMemberName_V` never existed in an committed code. It's OK to remove the `LINK_TO_STATIC_ARGS_V` as part of this commit of you wish.